### ISSUE TITLE
linter: Incorrect `no-unassigned-vars` lint with assignment assertion

### ISSUE THREAD

**Issue by tealsnow** [oxc-project/oxc#14621]
The typescript syntax:
```ts
let some_var!: SomeType;
```
Exists to assert that this variable *will* be assigned before use.
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-7.html#definite-assignment-assertions

But the linter incorrectly warns about the assignment with `no-unassigned-vars`, claiming it is never assigned. Again, I stress, the whole point of the syntax is to assert to the typescript compiler that it *will* be assigned in a way that the compiler cannot figure out itself.

This is used in particular with solidjs for refs:
```ts
let ref!: HTMLDivElement;

onMount(() => {
    // can use ref without errors, since it will be assigned by the time it is used
})

return <div ref={ref}>
   ...
</div>
```

**Comment by camc314** [comment]
Interestingly we have the same behaviour as eslint here:

https://eslint.org/play/#eyJ0ZXh0IjoibGV0IHJlZlxuXG5vbk1vdW50KCgpID0+IHtcbiAgICAvLyBjYW4gdXNlIHJlZiB3aXRob3V0IGVycm9ycywgc2luY2UgaXQgd2lsbCBiZSBhc3NpZ25lZCBieSB0aGUgdGltZSBpdCBpcyB1c2VkXG4gIHJlZjtcbn0pXG5cblxuY29uc3QgeCA9IDxkaXYgcmVmPXtyZWZ9IC8+Iiwib3B0aW9ucyI6eyJydWxlcyI6eyJuby11bmFzc2lnbmVkLXZhcnMiOlsiZXJyb3IiXX0sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6eyJqc3giOnRydWV9fSwiZWNtYVZlcnNpb24iOiJsYXRlc3QifX19

ESLint ref: https://github.com/eslint/eslint/issues/20169

**Comment by lilnasy** [comment]
... and typescript-eslint as well

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=JYOwLgpgTgZghgYwgAgBrIN7IL4FgBQANhGMgB4CEAXGgQTAK4gJjAD2IyMAFAJSbICyZFBIMonMgTz4CQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaWXAQ2WQEsBzXRAE0IDdTpUNJFpp9pJwwAXxCCgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

Though we could probably do better than these two do in isolation because oxlint rules have access to TypeScript's AST.

<img width="857" height="194" alt="Image" src="https://github.com/user-attachments/assets/0f8ad4a5-12cf-4d21-9532-b6fe5aedcf6a" />

**Comment by tealsnow** [comment]
Huh, I never got this issue with eslint. It only showed up when I tried to migrate from it to oxlint.

**Comment by lilnasy** [comment]
@camc314 Should we support this?

**Comment by camc314** [comment]
> [@camc314](https://github.com/camc314) Should we support this?

I actually haven't decided. I think the problem with supporting typescript syntax, is that adding the non-null assertion is firstly, a very small character, and the purpose of a linter is to try and catch correctness issues. 
For example, you could have code like:
```
let t!: string
```

that is really never assigned, but since the non-null assertion operator is there, it wouldn't be reported.

I'm leaning towards closing this as won't fix. This would have the same behaviour as eslint, biome, and doesn't rely on the internals of the solidjs compiler.

That said, i am open to other thoughts

**Comment by lilnasy** [comment]
I think if oxlint is going to be integrated with tsgo, we have to support this eventually. It would be unexpected if `oxlint --type-aware` complains about unassigned variables with the assertion. Though this bug could be categorized differently in that case.

**Comment by camc314** [comment]
tsgolint would be a slightly different case. we need to compile the svelte to pure JS/TSX first, and then run tsgolint on it. at the point that it's been compiled `ref` is correctly assigned a value so it wouldn't be incorrectly marked as unused.

**Comment by lilnasy** [comment]
Interesting. I tried with solid's compiled output, and it isn't an error there either, which is good. But wouldn't tsgolint act on the source code in solid's case since it's a custom JSX transform?

**Comment by camc314** [comment]
> But wouldn't tsgolint act on the source code in solid's case since it's a custom JSX transform?

oh solid might be a bit of a different case since it's already pure JSX, it would need more exploration. But for the other templating languages, vue for example, we would map forward/back so this wouldn't be an issue

**Comment by jamesarosen** [comment]
An alternative would be to support `argsIgnorePattern` just like `no-unused-vars` does.

Solid developers could set `argsIgnorePattern: "Ref$"` and name their refs with a `Ref` suffix.
